extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -694,4 +742,95 @@ export class AmpAnalytics extends AMP.BaseElement {
   }
 }
 
+/**
+ * Keeps track of the third-party vendors' cross-domain iframes. Only one
+ * static instance will be created, which will keep and manage ALL amp-analytics
+ * tags.
+ */
+class AmpAnalyticsFrames {
+  constructor() {
+    this.requestsFrameMap = map();
+  }
+
+  /**
+   * Create a cross-domain iframe
+   * @param {!HTMLDocument} parent  The document node of the parent page
+   * @param {!string} requestsFrameUrl  The URL, including data hash if
+   * applicable, of the cross-domain iframe
+   * @return {!Element}
+   */
+  create(parent, requestsFrameUrl) {
+    // Warning: the scriptSrc URL below is only temporary. Don't check
+    // in before getting resolution on that.
+    const frame = createElementWithAttributes(parent, 'iframe', {
+      sandbox: 'allow-scripts',
+      name: JSON.stringify({
+        'scriptSrc': '/examples/analytics-3p-remote-frame-helper.js',
+      }),
+    });
+    loadPromise(frame).then(() => {
+      this.setIsReady(requestsFrameUrl);
+    });
+    frame.src = requestsFrameUrl; // Set AFTER loadPromise to avoid race
+    setStyles(frame, {width: 0,height: 0,visibility: 'hidden'});
+    this.requestsFrameMap[requestsFrameUrl] = {
+      frame,
+      isReady: false,
+      msgQueue: [],
+    };
+    return frame;
+  }
+
+  /**
+   * Returns whether a requestsFrameUrl is already known
+   * @param {!string} requestsFrameUrl
+   * @return {!boolean}
+   */
+  has(requestsFrameUrl) {
+    return hasOwn(this.requestsFrameMap, requestsFrameUrl);
+  }
+
+  /**
+   * Sends a message to a cross-domain frame, or queues the message if the
+   * frame is not yet ready to receive messages.
+   * TODO: Implement throttling if messages are sent too rapidly.
+   * @param {!string} requestsFrameUrl
+   * @param {!string} message
+   */
+  send(requestsFrameUrl, message) {
+    const requestsFrame = this.requestsFrameMap[requestsFrameUrl];
+    if (requestsFrame.isReady) {
+      this.send_(requestsFrame.frame, [message]);
+    } else {
+      requestsFrame.msgQueue.push(message);
+    }
+  }
+
+  /**
+   * Send an array of messages to a cross-domain iframe
+   * @param {!Element} frame  The cross-domain iframe
+   * @param {!Array<string>} messages  The messages to send
+   * @private
+   */
+  send_(frame, messages) {
+    // Warning: the following code is likely only temporary. Don't check
+    // in before getting resolution on that.
+    frame && frame.contentWindow &&
+      frame.contentWindow.postMessage({ampAnalyticsEvents: messages}, '*');
```

Comment:
I assume all browsers we support will handle passing content as an object instead of string?  Is this what we do in window context?  Also, I thought we wanted to share the window.context infrastructure?